src/satipc.c

```diff
@@ -617,51 +618,120 @@ int satip_standby_device(adapter *ad) {
+    int i, rr, num_msg = 0, size_msg = 0;
+    int force_ok = 0;
+    struct mmsghdr messages[MAX_RTP_MSG] = {0};
+    struct iovec iovs[(MAX_RTP_MSG * 2) + 1] = {0};  // RTP Header + Payload (up to 1316 bytes) for each UDP datagram
     uint16_t seq;
     adapter *ad;
     satipc *sip;
     get_ad_and_sipr(ss->sid, 0);
-    struct iovec iov[3] = {{.iov_base = buf1, .iov_len = 12},
-                           {.iov_base = buf, .iov_len = len},
-                           {NULL, 0}};
-    *rb = readv(socket, iov, 2); // stripping rtp header
-    if (*rb > 0) {
-        ad = get_adapter(ss->sid);
-        sip->rcvp++;
-
-        copy16r(seq, buf1, 2);
-        if (sip->rtp_seq == 0xFFFF)
-            sip->rtp_seq = seq;
-        if (seq > sip->rtp_seq)
-            sip->rtp_miss++;
-        else if (seq < sip->rtp_seq)
-            sip->rtp_ooo++;
-        sip->rtp_seq = (seq + 1) & 0xFFFF;
+
+    // Prepare receiving multibuffer
+    for (i = 0; i < MAX_RTP_MSG && size_msg < len; i++) {
+        num_msg++;
+        size_msg += 1316;
+        if (size_msg > len) {  // Insufficient space to receive a 7*188 Bytes TS packet
+            size_msg -= 1316;
+            num_msg--;
+            break;
+        }
+
+        struct iovec *iov = &iovs[i * 2];   // stripping out RTP header, to leave continous payload data over the same buffer
+        iov[0].iov_base = buf1 + (i * 20);
+        iov[0].iov_len  = 12;
+        iov[1].iov_base = buf  + (i * 1316);
+        iov[1].iov_len  = (size_msg % 1316) ? : 1316;
+        iov[2].iov_base = NULL;
+        iov[2].iov_len  = 0;
+
+        struct mmsghdr *msg = &messages[i];
+        msg->msg_hdr.msg_iov = iov;
+        msg->msg_hdr.msg_iovlen = 2;
     }
-    if (!ad)
-        ad = get_adapter(ss->sid);
-
-    // Workaround for some poor SAT>IP servers (ex. XORO).
-    //      Some servers send KEEP-ALIVE RTP packets without valid TS payload.
-    //      Then RTP packets received without valid TS data are discarded.
-    uint8_t *b = buf;
-    if (*rb > 12 && *rb < DVB_FRAME - 12 && b[0] != 0x47) {
-        LOGM("discarding RTP packet without valid TS payload (sock %d, "
-             "socket_id "
-             "%d) [len=%d]",
-             socket, ss->id, *rb - 12);
-        *rb = 0;
-        return 1;
+
+    *rb = 0;
+
+    if (size_msg <= 0) {
+        LOG("satipc_read: Error allocating memory!")
+        return 0;
     }
 
-    if (ad && sip->ignore_packets) {
-        *rb = 0;
-        return 1;
+    rr = recvmmsg(socket, messages, num_msg, MSG_DONTWAIT, NULL);
+    if (rr <= 0) {
+        DEBUGM("satipc_read: read error or zero datagrams")
+        return 0;
     }
-    *rb -= 12;
-    return (*rb >= 0);
+
+    size_msg -= (num_msg - rr) * 1316;  // update the theoretical total size of the read buffer
+
+    // Loop over all datagrams received
+    ad = get_adapter(ss->sid);
+    uint8_t *bf1 = buf1;
+    for (i = rr - 1; i >= 0; i--) {  // loop in reverse order (i = 0; i < rr; i++)
```

Comment:
Why looping in reverse order?
Will the rtp sequence number checking work in this case?

Reply:
> Why looping in reverse order?

To be more efficient. When reading multiple RTP packets and pushing the PAYLOAD directly in the READ BUFFER, you can obtain HOLES in the buffer. So, if you're reading in reverse order then you only move valid data, and never any hole. The final result is identical, but in reverse order you copy less data if the number of holes is greater than one.

As a note, please take in consideration that a hole is when a RTP packet has a payload of less than 1316 bytes (that's less than 7 TS packets).

> Will the rtp sequence number checking work in this case?

Opps! In this point you're right. I need to revise this. Thank you for the tip.

Reply:
Hi @catalinii ,

This is solved. Now processed in forward way.